### ISSUE TITLE
ADirEx round 2 add In File search - regex and a new feature, Scan an existing cursor of files and do in file search on them.

### ISSUE THREAD

**Issue by myearwood1** [ChristianEhlscheid/vfp2c32#43]
Hi Christian

I wanted to use ADirEx to help GoFish speed up. That didn't work as I hoped. I ended up showing/helping integrate cygwin grep.exe. Grep is fast. The problem is we cannot get grep to process a large list of files. It can recurse folders, but not a large list of specific files.

Doing so in FoxPro has too much overhead from using FileToStr to import a file then RegEx it. My little DirX program proves your code is super fast. Therefore it stands to reason that if you could combine AdirEx with RegEx while scanning for filespec(s) in a folder hierarchy, and we use your callback functions, we might have a superior search engine for GoFish which is why I was looking at your utilities in the first place. It never even crossed my mind to ask you to combine a search within the file ability into ADirEx, but that should be the fastest way possible. GoFish supports plain text search, text search with wildcards and regex search.

Additionally, would you provide a new scan feature to accept an existing cursor of file names and search each of them?

This would allow a dev to find all files meeting a first criteria with ADirEx, then with ScanEx, search the ADirEx cursor/table/array for files with matching content to a target cursor and then ScanEx that cursor and so on.

To paraphrase Star Wars, Help us Chris-ti-an, you're our only hope. Chris-ti-an Kenobi :)

I'll test it for you.

Mike

**Comment by ChristianEhlscheid** [comment]
Hello Mike,

I didn't have any spare time in the last weeks because my father is severly ill.

I've coded something in the last days that will most probably do the job, almost exclusively foxpro code, asyncronous file IO combined with a freeware regex C library.
I'm hesitent to include this into the fll itself since this is just too specific.
But I'm certain the FoxPro code will fit the bill.

I'll contact you again when I have a working sample in the coming days.

Regards
Christian

**Comment by myearwood1** [comment]
Hi Christian.

I'm sorry to learn of your father's illness. I hope he recovers. I've read up a lot on these regex things. I am leaning towards PCRE2.

Thanks

Mike

**Comment by myearwood1** [comment]
I have been trying to create a small fll that a FoxPro developer could use without needing filetostr to read the contents, but allow the c++ function to do it. It will also be able to regex a memo field or memvar. I expect that will be faster than existing FoxPro methods.

**Comment by ChristianEhlscheid** [comment]
Hello Mike,

there will probably be hardly any difference in runtime, but FILETOSTR is not the right tool for sure.

The key on modern storage system's (SSD's) is to generate a ton of small reads in a short time asynchronously and let the OS handle all the heavy lifting to bring the data into memory. Syncronous IO just kills performance. This can also be done from within FoxPro.

The code I've written was able to search 275.000 PHP source files in 2 seconds, and there are still opportunities for more aggressive optimization, I hope thats fast enough.

I'm almost finished, I will post the code later today.

**Comment by ChristianEhlscheid** [comment]
Here it is ...

[regex_search.zip](https://github.com/user-attachments/files/16089478/regex_search.zip)

The regex project I've used is located here: [TRE Regex](https://github.com/laurikari/tre)

**Comment by myearwood1** [comment]
Hi Christian - very interesting. I like that it says "TRE attempts to conform to these specifications as strictly as possible. " I have stubbed my toe with the regex used in Gofish in the recent past, FindStr, cygwin Grep Perl version has some bugs. I will study your code with great interest! 
Thanks!

**Comment by myearwood1** [comment]
I had to give it a try before I head home! 
Right after the TRY in the init of regex_t there is the line THIS.Address = AllocMem(THIS.SizeOf) which throws an exception on line 338. The allocation attempt failed because of heap corruption or improper function parameters. THIS.SizeOf is 8 as set just a few lines before. I have your very latest vfp2c32.fll which is where it's getting the AllocMem. I also have vfp2c.h, your windows.h the tre*.dll in a separate folder and no FoxPro path for it to find stuff.

**Comment by ChristianEhlscheid** [comment]
Hello Mike,

I've fixed a couple bugs and made some other optimizations.
The problem you encountered was probably a 64bit issue, if not let me know.

Here's the new version 
[regex_search.zip](https://github.com/user-attachments/files/16098721/regex_search.zip)

**Comment by myearwood1** [comment]
Respected sir.

I appreciate you taking the time. I did not anticipate a quick response due to your situation. I have not heard from any relatives in St. Vincent since the hurricane. I'll let you know once I've tried it.

Thanks

Mike

**Comment by myearwood1** [comment]
Hi Christian

THIS IS BRILLIANT! Awesome code. You seem to have broken each file into multiple packets and assembled them before calling ProcessFileIO. 

I got a different error number but still crashed in the regex_t init, until I did more digging. There is a difference between the older vfp2c32.fll and the more recent one I got from your site. For one thing, the recursion in the newer adirex did not work, nor the fullpathnames. The previous vfp2c32.fll works. I put the older vfp2c32 in vfp2c1 and the newer in vfp2c2. Then it got passed the error THIS.Address = AllocMem(THIS.SizeOf) in regex_t.init.

It carried on a bit until it reached this line:

m.loAsyncFileReader.hFile = CreateFileA(m.lcFileName, GENERIC_READ, FILE_SHARE_READ, 0, 3, FILE_FLAG_OVERLAPPED, 0)

and gave the error message variable FILE_FLAG_OVERLAPPED is not found. I found that should be 0x40000000, so plugged that in and continued. I'm familiar with using CreateFile to Open an existing file.

I think I'll add a flag to searchfile to have it stop on the first match and have it return a t/f if it detected anything. Then make processfileio return a t/f also, so I can react to a match. If that was done in C++ vs VFP code, c++ would win though, right?

Bravo Maestro

**Comment by ChristianEhlscheid** [comment]
Hello Mike,

C/C++ would be faster for sure, how much is hard to predict, maybe a factor of 2 or 3 at most. Since at the end of the day the storage media (SSD/HDD) random read speed is probably the limiting factor. With the completion port IO used it's also possible to spawn several threads that wait on finished async IO with GetQueuedCompletionStatus and do the regex searching in multiple threads. I don't know how much optimization effort you want to put into this, For the size of projects I've worked on with a couple hundred of source files at most, how it performs currently is probably sufficient for most developers.

Searching all the forms and classlibs will be more time consuming than searching .prg files. When you get to that part you may contact me again, i've had some ideas to optimize this task.

**Comment by myearwood1** [comment]
Hi Christian 

The app I inherited has over 20,000 files. We are lucky to have source safe converted versions of all files. So that's down to around (1121 .prg files and 5470 .??a files) 6000 files. Gofish was not using a locate command to run the regex, which I was pretty sure of, so they fixed that.

I then got gofish to reluctantly use grep, but grep has limitations, one being it cannot read a file listing files to process.

For a search utility that seems shortsighted. If it could use wildcards to scan the directory, then put the matching files in a text file, then a user could process the subset and so on.

Once the list of files is found, gofish then has to scan the scx, vcx directly using regex to show the matches as they have done for years.

The project I inherited is admittedly huge, it grew like weeds like so many systems so every part of gofish was too slow for me, but no one seems to be concerned as long as the mob is happy.

Gofish is not using grep to search in the second pass, which is a problem.

My wish is for a c++ routine that can regex files or memo fields or memory variables. Gofish and the larger Fox community can use such a thing. It would be nice to have a reliable regex utility available to all.

**Comment by myearwood1** [comment]
How is your father? My mother got Dengue fever from the increased mosquitoes from the hurricane. She's OK now.

**Comment by ChristianEhlscheid** [comment]
Hello Mike,

thanks for asking, sadly my father passed away last tuesday. 

I already started work on writing a FLL wrapper for Intel's Hyperscan, probably the fastest regex library available, usually used in realtime network security software for scanning network traffic. 

At the moment I'm dealing with the consequences of my fathers death, but I will surely finish this in the future.

Please be patient a few weeks, I'm confident I can release something at the end of the month.
It looks very promising, the current version is really capable to scan about 30000 source files in 1 second.
It will scale with the number of cores in the system using the windows threadpool api and asyncronous IO.

The current botlleneck is the AdirEx function itself, so I plan the optimize this aswell, I've read up on alternatives to the FindFirstFileEx/FindNextFile api, and there exists lower level api functions that I can use that should speed it up considerably.

Regards
Christian

**Comment by myearwood1** [comment]
Honored sir, you have my sympathies.  Please take your time. I know this will be amazing.

**Comment by myearwood1** [comment]
Happy New Year Christian!

I hope you are well. I just read about something and thought of you!

https://github.com/ned14/llfio

**Comment by ChristianEhlscheid** [comment]
Hello Mike,

it has taken way longer than I anticipated, but I've just released a new version with an improved AdirEx function and
with an integration of Intel's hyperscan regex library. 
If you're still in need of this functionality you might have a look.

**Comment by myearwood1** [comment]
Hi Christian

That is great of you doing it. Can you email me directly?
***@***.***

I have news.

Mike
